Leave invisible characters in string fields

garethbowen opened the pull request:
# Description

This is a further improvement to stripping invisible characters. Instead of always
removing invisible unicode characters, now they're only removed from non-string
fields so we don't modify values where the character may be desired.

medic/cht-core#7676

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

garethbowen commented:
@binokaryg Can I get a review for this work? There's a bit of refactoring but hopefully it's the best of both worlds for processing zero width characters.

binokaryg reviewed (approved):
Thanks for working on this improvement. It looks better.

Few things that I have noticed:
- To benefit from this, we need to make sure that the filed Patient ID should be an integer type, not a string in the configuration.
- Ther zero-width characters are preserved in the raw message (`doc.sms_message.message`), which I think is good. This way, we can know when the users are typing these characters, if we want to. However, in the raw message too, the zero-width character is removed from the form code. Could there be an easy way to preserve it?

garethbowen commented on api/tests/mocha/services/records.spec.js:
```diff
@@ -54,14 +54,22 @@ describe('records service', () => {
+    const formDefinition = 'YY\u200BYY'; // strip the invisible character to match form defn
+    const facilityId = 'Faci\u200Clity'; // string value - do not strip
+    const year = '19\u200C99'; // integer value - strip
+    const month = '1\u200D2'; // enum value - strip to match
```
Added an additional assertion to check this.

garethbowen commented:
> we need to make sure that the filed Patient ID should be an integer type, not a string in the configuration.

I don't think that'll work because the patient ID may start with a leading 0 (eg: "01234") which would get truncated if you convert to an integer. I've added special handling for `patient_id` and `place_id` which I think works.

> However, in the raw message too, the zero-width character is removed from the form code. Could there be an easy way to preserve it?

It was a bit tricky because of the spaghetti like nature of this code but I've added a line to reset it to the raw content before saving it which I think achieves this, albeit in a hacky way.

garethbowen commented:
@binokaryg Thanks! I've made those improvements - can you have another look please?

binokaryg reviewed (approved):
Thanks for the improvements.
